### PR TITLE
core(fr): add navigation phase + trace gatherer

### PR DESCRIPTION
**Pull request by patrickhulce**
**Summary**
The primary goal of this PR is to add the trace gatherer and support performance metrics in Fraggle Rock navigation mode (as well as pave the way for upgrades to support timespan metrics). That change is pretty small, but this triggered a number of other changes.

- Add new `beforeNavigation`/`afterNavigation` methods to distinguish this new category of "navigation-only" gatherers.
- Adjust config validation rules to disallow navigation gatherers from depending on snapshot artifacts.
- Refactor navigation runner to reduce duplication in collection of artifacts of each phase (there are now 5 instead of just 3)
- Refactor default trace categories list to new gatherer, import from legacy driver.
- Add more test coverage for these refactored areas.

**Primary Point of Discussion**

This PR sets the new execution sequence to the following phase order. 

- beforeNavigation
- beforeTimespan
- afterTimespan
- afterNavigation
- snapshot

The primary question IMO is when exactly `beforeNavigation` and `afterNavigation` should take place.

There is an argument for moving `afterNavigation` to happen *after* `snapshot` so it can depend on snapshot artifacts, but that means any timespan style recording would include all of the snapshot gatherer work which seems undesirable. I've kept it before `snapshot` and if anything truly needs the `snapshot` dependency we can think of something like `afterSnapshot` to fill that gap in the future.

**Implications for Gatherer Authoring**

There are 5 potential phases, and this is where that array of supported modes starts enter the territory that concerned me. Really we only have 3 types of valid gatherers. Timespan, Snapshot, or Navigation. We can enforce that fact in config validation if we like. The gatherer of each type should *only* truly implement the set of methods that are related to it *but* all the methods of the supported modes will be invoked. 

Example: A timespan gatherer says it works for both timespans and navigations, but it only implements `beforeTimespan` and `afterTimespan`, leaving the rest as noops. When run in navigation mode, `beforeNavigation` and `afterNavigation` will still be executed.

**Related Issues/PRs**
ref #11313

### CODE REVIEW

File: lighthouse-core/fraggle-rock/gather/base-gatherer.js [hunk 1]
```diff
@@ -20,11 +20,11 @@ class FRGatherer {
   meta = {supportedModes: []}
 
   /**
-   * Method to gather results about a page in a particular state.
+   * Method to start observing a page before a navigation.
    * @param {LH.Gatherer.FRTransitionalContext} passContext
-   * @return {LH.Gatherer.PhaseResult}
+   * @return {Promise<void>|void}
    */
-  snapshot(passContext) { }
```
Comment (patrickhulce) [inline lighthouse-core/fraggle-rock/gather/base-gatherer.js:27]: moved this just to reflect the order in which they're invoked

File: lighthouse-core/gather/driver.js [hunk 2]
```diff
@@ -139,43 +137,7 @@ class Driver {
   }
 
   static get traceCategories() {
```
Comment (patrickhulce) [inline lighthouse-core/gather/driver.js:139]: this is part of our public API, so didn't remove

File: lighthouse-core/gather/driver.js [hunk 3]
```diff
@@ -816,15 +778,7 @@ class Driver {
   async beginTrace(settings) {
     const additionalCategories = (settings && settings.additionalTraceCategories &&
         settings.additionalTraceCategories.split(',')) || [];
-    const traceCategories = this._traceCategories.concat(additionalCategories);
-
-    // In Chrome <71, gotta use the chatty 'toplevel' cat instead of our own.
```
Comment (patrickhulce) [inline lighthouse-core/gather/driver.js:821]: with Chrome stable up to 88 this feels pretty safe now :)

File: lighthouse-core/gather/gatherers/trace-compat.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+const TraceGatherer = require('./trace.js');
+const FRGatherer = require('../../fraggle-rock/gather/base-gatherer.js');
+
+/** @implements {LH.Gatherer.FRGathererInstance<'Trace'>} */
+class TraceCompat extends FRGatherer {
+  /** @type {LH.Gatherer.GathererMeta<'Trace'>} */
+  meta = {
+    supportedModes: ['navigation'],
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/trace-compat.js:21]: this is actually a fairly important signal that we shouldn't try to compute all the metrics over an arbitrary timespan

longer term I think this evolves into a `Trace` and `NavigationTrace` artifact or something where the latter is guaranteed to have been collected over a navigation

File: lighthouse-core/test/fraggle-rock/config/config-test.js [hunk 5]
```diff
@@ -230,8 +230,23 @@ describe('Fraggle Rock Config', () => {
...
+    });
+
+    it('should throw when timespan needs navigation', () => {
+      dependentGatherer.meta.supportedModes = ['timespan'];
+      dependencyGatherer.meta.supportedModes = ['navigation'];
+      expect(() => initializeConfig(configJson, {gatherMode: 'navigation'}))
+        .toThrow(/Dependency.*is invalid/);
+    });
+
+    it('should throw when navigation needs snapshot', () => {
```
Comment (patrickhulce) [inline lighthouse-core/test/fraggle-rock/config/config-test.js:247]: complete coverage is already in validation-test but covering some more important common mistakes seemed useful

File: types/gatherer.d.ts [hunk 6]
```diff
@@ -100,9 +100,11 @@ declare global {
     export interface FRGathererInstance<TDependencies extends DependencyKey = DefaultDependenciesKey> {
       name: keyof LH.GathererArtifacts; // temporary COMPAT measure until artifact config support is available
       meta: GathererMeta<TDependencies>;
-      snapshot(context: FRTransitionalContext<TDependencies>): PhaseResult;
```
Comment (patrickhulce) [inline types/gatherer.d.ts:103]: rearranged to signal the order in which they're invoked

File: lighthouse-core/gather/driver.js (hunk 2 above)
Comment (patrickhulce) [reply lighthouse-core/gather/driver.js:139]: ya let's do it

Comment (patrickhulce) [inline lighthouse-core/gather/driver.js:139]:
```suggestion
  /** @deprecated - Not available on Fraggle Rock driver. */
  static get traceCategories() {
```

**Review by adamraine — APPROVED** [review]

**Comment by patrickhulce** [comment]
I take it there are no strong feelings about the order of phase execution yet. I wouldn't really expect there to be without hands-on experience yet, so let's leave this as a good first start and if we run into any issues while converting the remaining gatherers we can revisit the decision if necessary.

**Pull request merged**